### PR TITLE
Add distributed lock implementation for Dynamo DB

### PR DESCRIPTION
**Pull request by dhickie**
This adds a DynamoDB implementation of the distributed lock provider used by the outbox sweeper and archiver.

### CODE REVIEW

File: src/Paramore.Brighter.Locking.DynamoDB/DynamoDbLockingProvider.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            if (_options.ManuallyReleaseLock)
+            {
+                try
+                {
+                    await _dynamoDb.DeleteItemAsync(BuildReleaseRequest(resource, lockId), cancellationToken);
+                }
+                catch (ConditionalCheckFailedException)
+                {
+                    // Log that the lease is no longer valid
```
Comment (iancooper) [inline src/Paramore.Brighter.Locking.DynamoDB/DynamoDbLockingProvider.cs:85]: Generally we have a model of creating a static logger in the class, and then using that. See: https://github.com/BrighterCommand/Brighter/blob/master/src/Paramore.Brighter.Locking.Azure/AzureBlobLockingProvider.cs

Comment (dhickie) [reply src/Paramore.Brighter.Locking.DynamoDB/DynamoDbLockingProvider.cs:85]: Ah yep, I completely missed the logging. I'll sort that out.

**Comment by iancooper** [comment]
No worries. Forgetting stuff, is what reviews can help with.

    On Thursday, 18 July 2024 at 14:50:57 BST, dhickie ***@***.***> wrote:  
 
 


@dhickie commented on this pull request.

In src/Paramore.Brighter.Locking.DynamoDB/DynamoDbLockingProvider.cs:
> +        /// </summary>
+        /// <param name="resource">The name of the resource to Lock</param>
+        /// <param name="lockId">The lock Id that was provided when the lock was obtained</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>Awaitable Task</returns>
+        public async Task ReleaseLockAsync(string resource, string lockId, CancellationToken cancellationToken = default)
+        {
+            if (_options.ManuallyReleaseLock)
+            {
+                try
+                {
+                    await _dynamoDb.DeleteItemAsync(BuildReleaseRequest(resource, lockId), cancellationToken);
+                }
+                catch (ConditionalCheckFailedException)
+                {
+                    // Log that the lease is no longer valid

Ah yep, I completely missed the logging. I'll sort that out.

**Review by iancooper — APPROVED** [review]

**Comment by iancooper** [comment]
@dhickie I'll merge, because the broken tests are known (aws ci and issues with the posgres lock)

**Pull request merged**